keps/sig-instrumentation/3077-contextual-logging/README.md

```diff
@@ -0,0 +1,974 @@
+
+### Troubleshooting
+
+###### How does this feature react if the API server and/or etcd is unavailable?
+
+Works normally.
+
+###### What are other known failure modes?
+
+None besides bugs that could cause a program to panic (null logger).
+
+###### What steps should be taken if SLOs are not being met to determine the problem?
+
+Revert commits that changed log calls.
+
+## Implementation History
+
+## Drawbacks
+
+Supporting contextual logging is a key design decision that has implications
+for all packages in Kubernetes. They don’t have to be converted all at once,
+but eventually they should be for the sake of completeness. This may depend on
+API changes.
+
+The overhead for the project in terms of PRs that need to be reviewed can be
+minimized by combining the conversion to contextual logging with the conversion
+to structured logging because both need to rewrite the same log calls.
+
+## Alternatives
+
+### Per-component logger
+
+A logger could be set for object instances and then all methods of that object
+could use that logger. This approach is sufficient to get rid of the global
+logger and thus for the testing use case. It has the advantage that log
+messages can be associated with the object that emits them.
+
+The disadvantage is that associating the log message with the call chain via
+multiple `WithName` calls becomes impossible (mutually exclusive designs).
+
+Enriching log messages with additional values from the call chain’s context is
+an unsolved problem. A proposal for passing a context to the logger and then
+letting the logger extract additional values was discussed in
+https://github.com/go-logr/logr/issues/116. Such an approach is problematic
+because it increases coupling between unrelated components, doesn’t work for
+code which uses the current logr API, and cannot handle values that weren’t
+attached to a context.
+
+Finally, a decision on how to pass a logger instance into stand-alone functions
+is still needed.
+
+### Propagating a logger to init code
+
+controller-runtime handles the case of init functions retrieving a logger and
+keeping that copy for later logging by handing out [a
+proxy](https://github.com/kubernetes-sigs/controller-runtime/blob/78ce10e2ebad9205eff8429c3f0556788d680c27/pkg/log/deleg.go).
+
+This has additional overhead (mutex locking, additional function calls for each
+log message). Initialization of log output for individual test cases in a unit
+test cannot be done this way.
+
+It's better to avoid doing anything with logging entirely in init code.
+
+### Panic when DefaultLogger is called too early
+
+This would provide an even more obvious hint that the program isn’t working as
+intended. However, the log call which triggers that might not always be
+executed during program startup, which would cause problems when it occurs in
+production. Emitting a `FIXME` log message seems more suitable.
+
+## Infrastructure Needed
+
+A new staging repo `k8s.io/klogr` is needed.
```

Comment:
a staging repo or a standalone repo? do we expect projects outside k8s.io/kubernetes and staging repos to import and use this repo? if so, that means those projects can never be dependencies of anything within k8s.io/kubernetes or the staging repos.

Reply:
I prefer to start with a staging repo because then it'll be easier to modify the implementation together with Kubernetes, if necessary. We also have some more wiggle room for not committing to a stable API yet. Once we have a stable v1 API, it can become a standalone repo with its own versioning and releases.

Outside projects and other staging repos will use `k8s.io/klogr`. Therefore its own dependencies will have to be very minimal. Right now those are just go-logr, the Go standard packages and (for the initial transition period) klog.